Entities/AtomGroup.cpp

```diff
@@ -1,2618 +1,1750 @@
+		for (const Atom *intersectingAtom : intersectingAtoms) {
+			exitDirection += m_OwnerMO->RotateOffset(intersectingAtom->GetNormal());
+		}
+
+		// We don't have a direction to go, so quit
+		// TODO: Maybe use previous position to create an exit direction instead then?
+		if (exitDirection.IsZero()) {
+			return false;
+		}
+
+		// Invert and set appropriate length
+		exitDirection = -exitDirection;
+		exitDirection.SetMagnitude(m_OwnerMO->GetDiameter());
+
+		Vector clearPos = Vector();
+
+		// See which of the intersecting Atoms has the longest to travel along the exit direction before it clears
+		float longestDistance = 0.0F;
+		for (const Atom *intersectingAtom : intersectingAtoms) {
+			atomPos = intersectingAtom->GetCurrentPos();
+			if (g_SceneMan.CastFindMORay(atomPos, exitDirection, g_NoMOID, clearPos, 0, true, 0)) {
+				// Determine the longest clearing distance so far
+				atomExitVector = clearPos - atomPos.GetFloored();
+				if (atomExitVector.GetMagnitude() > longestDistance) {
+					// We found the Atom with the longest to travel along the exit direction to clear, so that's the distance to move the whole object to clear all its Atoms.
+					longestDistance = atomExitVector.GetMagnitude();
+					totalExitVector = atomExitVector;
+				}
+			}
+		}
+
+		// The final exit movement vectors for this' Owner MO and the MO intersected by this
+		Vector thisExit;
+		Vector intersectedExit;
+
+		// If the other object is pinned, then only move this. Otherwise, apply the object exit vector to both this owner object, and the one it intersected, proportional to their masses.
+		if (intersectedMO->GetPinStrength() > 0.0F) {
+			thisExit = totalExitVector;
+		} else {
+			float massA = m_OwnerMO->GetMass();
+			float massB = intersectedMO->GetMass();
+			float invMassA = 1.0F / massA;
+			float invMassB = 1.0F / massB;
+			float normMassA = invMassA / (invMassA + invMassB);
+			float normMassB = invMassB / (invMassA + invMassB);
+
+			// If the intersected is much larger than this' MO, then only move this. Otherwise, apply the movements to both this and the intersected MO's, proportional to their respective masses.
+			if (normMassB < 0.33F) {
+				thisExit = totalExitVector;
+			} else {
+				thisExit = totalExitVector * normMassA;
+				intersectedExit = -totalExitVector * normMassB;
+			}
+		}
+
+		// Now actually apply the exit vectors to both, but only if the jump isn't too jarring
+		if (thisExit.GetMagnitude() < m_OwnerMO->GetRadius()) { position += thisExit; }
+		if (!intersectedExit.IsZero() && intersectedExit.GetMagnitude() < intersectedMO->GetRadius()) { intersectedMO->SetPos(intersectedMO->GetPos() + intersectedExit); }
+
+		// If we've been pushed into the terrain enough, just gib as the squashing effect
+		if (m_OwnerMO->CanBeSquished() && RatioInTerrain() > 0.75F) /* && totalExitVector.GetMagnitude() > m_OwnerMO->GetDiameter()) */ {
+			// Move back before gibbing so gibs don't end up inside terrain
```

Comment:
This comment about moving back is actually good cause it's not self-explanatory. The comment above is not good because it's just describing the code.